---
title: "Enhancing gene set enrichment using networks"
author: 
  - name: Michael Prummer
    affiliation: NEXUS Personalized Health Technologies, ETH Zurich, and Swiss Institute for Bioinformatics, Zurich, Switzerland.
abstract: Differential gene expression (DGE) studies often suffer from poor interpretability of their primary results, i.e., thousands of differentially expressed genes. This has led to the introduction of gene set analysis (GSA) methods that aim at identifying interpretable global effects by grouping genes into sets of common context, such as, molecular pathways, biological function or tissue localization. In practice, GSA often results in hundreds of differentially regulated gene sets. Gene sets are often regulated in a correlative fashion because they share many of their genes or they describe related processes. Using these kind of neighborhood information to construct networks of gene sets allows to identify highly connected sub-networks as well as poorly connected islands or singletons. We show here how topological information and other network features can be used to filter and prioritize gene sets in routine DGE studies. Community detection in combination with automatic labeling and the network representation of gene set clusters further constitute an appealing and intuitive visualization of GSA results. The RICHNET workflow described here does not require human intervention and can thus be conveniently incorporated in automated analysis pipelines.
keywords: differential gene expression analysis, gene set analysis, enrichment analysis, network analyis, GSEA.
bibliography: richnet.bib
csl: science.csl
output: BiocWorkflowTools::f1000_article
editor_options: 
  chunk_output_type: console
---

# Introduction

Interpretation of whole transcriptome differential expression studies is often difficult because the sheer volume of the differentially expressed genes (DEGs) can be overwhelming. It is common place in designed experiments with more than just a marginal biological effect to find several thousands of differentially expressed genes (DEGs). One way to handle the vast numbers and to identify the biological consequences of gene expression changes is to associate them with overarching processes involving a whole set of genes, such as, GO terms or KEGG pathways. 

Curated genesets have been designed or discovered for a wide range of common contexts, such as, a biological process, molecular pathway, or tissue localization [@Rouillard2016; @Liberzon2011]. They have been introduced in the past not only to reduce complexity and to improve interpretability but also to increase statistical power by reducing the number of performed tests. As it turns out, this often results in finding hundreds of differentially regulated pathways[^foot1]. 

[^foot1]: The terms _geneset_ and _pathway_ are used interchangeably throughout this document and refer to a set of genes.

As with co-expressed genes, many of the pathways exhibit strong mutual correlation because they contain a large proportion of shared genes which is in turn a result of the fact that many of them describe closely related aspects of an overarching biological theme. Therefore, to further increase interpretability of differential geneset regulation and to capture the global change of a biological phenotype, it would be desired to identify possibly existing umbrella organizations among genesets.

Networks are ideal to model dependencies, interactions, and similarities among individuals [@Barabasi2004; @Vidal2011; @Ideker2012], be it people, computers, genes, or genesets. The degree of connectivity between them can have an influence on information flow and defines communities or _cliques_, i.e., clusters of highly connected nodes within and infrequent connections between them.

In order to construct a geneset network a similariy measure is required and can be defined as the fraction of common genes, also called the Jaccard index [@Merico2010]. Other ways to measure similarity among genesets include, for instance, coexpression strength as implemented in WGCNA [@Langfelder2008; @Thorsson2018].

Community detection based on network topology is a standard problem in the analysis of social networks [@Girvan2002; @Bedi2016]. Well-established algorithms allow for computationally efficient clustering of genesets and can be used to identify highly connected sub-networks. There is no unique or optimal method available but many options exist. Popular methods to define clusters include the _edge-betweenness_ criterion, the _Infomap_ or the _Louvain_ algorithm (`igraph`) as well as hierarchical or kmeans clustering.
 
Once geneset clusters are defined they can be characterized by their size and connectivity and thus prioritized and ranked. In particular, the clusters can be categorized in singletons, doublets, medium and large or dense and loose clusters. 

Network analysis not only allows to detect clusters and perform measurements on them, networks are also straightforward and appealing visualizations of similarities among geneset. There are a couple of interactive visualization software tools available, of which Cytoscape is probably the most popular [@Shannon2003]. In some cases interactivity is useful but the emphasis here is to provide some of Cytoscape's features without any human intervention for easy integration into automatic analysis pipelines. For instance, automatic labeling of communities using the n most frequent terms was addopted here similar as in [@Kucera2016].

The purpose of this step-by-step workflow is to provide a fully automated and reproducible procedure for downstream analysis and visualization of differential geneset analysis results in R. The focus is on supporting scientists in result interpretation by bringing order into the list of differentially regulated genesets based on biological rather than pure statistical arguments. The workflow is suitable for any kind of geneset library including new or custom sets and any kind of geneset analysis method.

Starting with differential expression analysis of a model dataset, geneset analysis is performed based on the MSigDB library. A geneset network is constructed to identify isolated genesets (singletons) and geneset pairs (doublets). Larger connected sub-networks are then split into smaller clusters of closely related genesets describing similar processes. The effect of each modification step on the network topology is visually documented in Figs. 1-4. Using the most frequently occuring terms in the geneset names of a cluster, an attempt to autmatically assign cluster labels is made. Finally, all labeled clusters of genesets are plotted to provide a one page overview of the results (Fig. 5).


# Preparations

The packages required for this workflow provide plotting functions (`ggplot2` and relatives), network functions `igraph`[@Csardi2006] and `GGally`, text analytics functions (`wordcloud`, etc.) and gene expression analysis functions `DESeq2`[@Love2014], `limma`[@Ritchie2015], and `org.Hs.eg.db`.

```{r setup, include=F, echo=T}
knitr::opts_chunk$set(echo=T, error=F, message=F, warning=F, cache=T, results='asis', dpi=300, fig.show = 'asis')
```

```{r init, results="hide"}
lby = c("RColorBrewer", "ggplot2", "gplots", "cowplot", 
        "ggrepel", "reshape2", "knitr", "kableExtra",
        "igraph", "GGally", 
        "DESeq2", "limma", "org.Hs.eg.db", 
        "wordcloud", "tm", "SnowballC")
tmp = lapply(lby, require, character.only=T, warn.conflicts=F, quietly=T)
```

In addition to and often based on `igraph`, several R packages for network visualization are available and described in the form of tutorials [@Ognyanova2015; @Tyner2017]. 


### Example data

We are using the popular _airway_ data set [@Himes2014] and perform a simple differential expression analysis.

```{r dea}
library(airway)
data(airway)
dds = DESeqDataSetFromMatrix(countData = assay(airway),
                             colData = colData(airway),
                             design = ~ cell + dex)
dds$dex = relevel(dds$dex, "untrt")
dds = DESeq(dds, betaPrior = T)
res = results(dds, contrast = c("dex", "trt", "untrt"))
```

### Mapping Ensembl IDs to ENTREZ IDs

We are using the popular `org.Hs.eg.db` package based on the UCSC annotation database and keep only genes with a unique mapping.

```{r entrezmap, eval=T}
res$entrezgene = unname(mapIds(org.Hs.eg.db, keys = rownames(res), 
                               column = "ENTREZID", keytype = "ENSEMBL"))
res = subset(res, subset = !is.na(res$entrezgene) & !is.na(res$stat))
res = res[-which(duplicated(res$entrezgene)), ]
```


### Gene set enrichment analyis

We are using the popular KEGG, Reactome, and Biocarta pathways from the MSigDB gene set library C2. The following chunk guarantees that the gene set library list object is called `gset`.

```{r load_genesets}
url = "http://bioinf.wehi.edu.au/software/MSigDB/human_c2_v5p2.rdata"
temp.space = new.env()
bar = load(url(url), temp.space)
gset = get(bar, temp.space)
rm(temp.space)
gs.libs = sapply(names(gset), function(x) strsplit(x, "_")[[1]][1])
gset = gset[which(gs.libs %in% c("KEGG", "REACTOME", "BIOCARTA"))]
```

Competitive gene set enrichment analysis is performed using the function `camera()` from the `limma` package. We include uni-directional and bi-directional enrichment by using both the test statistics ("up" or "down") and its modulus ("mixed") for gene set testing. We limit the following network analysis to gene sets with a $FDR < 0.05$.

```{r gsa}
idx = ids2indices(gene.sets = gset, identifiers = res$entrezgene)
dat = cameraPR(res$stat, idx, sort = F)
dat$PValue.Mixed = cameraPR(abs(res$stat), idx, sort = F)$PValue
dat$FDR.Mixed = p.adjust(dat$PValue.Mixed, method = "BH")
dat$name = rownames(dat)

dat$Direction = as.character(dat$Direction)
dat$Direction[dat$FDR > 0.05] = "Mixed"
dat$Direction[dat$Direction == "Mixed" & dat$FDR.Mixed > 0.05] = "NOT"
dat$Direction = factor(dat$Direction, levels=c("NOT", "Up", "Down", "Mixed"))

idx = which(dat$Direction == "Mixed")
if(length(idx) > 0) dat$FDR[idx] = dat$FDR.Mixed[idx]
dat = dat[, -grep("\\.Mixed", names(dat))]
dat = dat[dat$Direction != "NOT", ]
dat$Direction = factor(dat$Direction, levels=c("Up", "Down", "Mixed"))
```

Starting from `r length(gset)` gene sets, `r nrow(dat)` are found to be differentially regulated. Many of them are expected to describe similar processes and to be highly correlated.


# Network construction

We construct a gene set network based on the proportion of common genes as the inverse distance measure. The nodes are gene sets which are connected by edges if the Jaccard index

$$ J = \frac{\text{Number of common genes}}{\text{Number of all genes}} $$

is larger than a preset threshold,  $J > 0.2$. While this threshold is somewhat arbitrary it has proven to be a reasonable one in many projects. Nevertheless, it is strongly recommended to investigate its effect on the quality  of the results. 

```{r adjacency}
# only keep gene sets present in the data
id.keep = which(names(gset) %in% dat$name)
gset = gset[id.keep]
# adjacency matrix
m.adj = sapply(gset, function(x) 
  sapply(gset, function(y) 
    length(intersect(unlist(x), unlist(y) ))
    )
  )
diag(m.adj) = 0
# Jaccard index matrix
NGenes = sapply(gset, length)
m.union = outer(NGenes, NGenes, "+") - m.adj
m.jacc = m.adj / m.union
```

The Jaccard matrix, or adjacency matrix, can be conveniently used to construct a network object using the function `igraph::graph_from_adjacency_matrix()`. In this example geneset similarity is measured using all member genes irrespective of whether they were detected and present in the data. Alternatively, one could include only genes present in the data depending on whether the current data seem more relevant and trustworthy or the prior information given by the geneset definition. Graphical display is achieved here using `ggnet::ggnet2()` (Figure \@ref(fig:figure1)).

```{r figure1, fig.cap="Graphical representation of the initial gene set network. Node colors indicate whether the member genes of a set are predominantly up or down regulated or whether there is no preferential direction (mixed).", fig.height=4, fig.width=5}
# choose node colors
palette = brewer.pal(9, "Set1")[c(1,2,9)]
names(palette) = c("Up", "Down", "Mixed")
# apply cutoff to Jaccard matrix
m.adj1 = m.adj * (m.jacc > 0.2)
# construct network object
net = graph_from_adjacency_matrix(m.adj1, "upper", diag = F, weighted = T)
# add vertex features
V(net)$size = dat$NGenes
V(net)$color = palette[dat$Direction]
V(net)$Direction = as.character(dat$Direction)
# plot
ggnet2(net, size = 2, color = "Direction", palette = palette, 
       edge.size = 1, edge.color = "#99CC33")
```

# Network modifications

In the following, components of the network for which network analysis does not improve interpretability are identified and put to aside. This includes singletons, i.e., genesets not connected to any other geneset, and doublets, also termed binary systems or dumbbells, i.e., pairs of genesets connected with each other but isolated from the rest.

### Identify singletons

```{r singletons}
singletons = which(igraph::degree(net) == 0)
net1 = delete_vertices(net, singletons)
in.single = which(dat$name %in% V(net)$name[singletons])
tab = dat[in.single, ]
tab$FDR = signif(tab$FDR, 2)
tab$name = gsub("_", " ", tab$name)
tab = kable(tab[,c("name", "NGenes", "Direction", "FDR")], 
            row.names = F, format = "latex", 
            caption = "List of all singletons, i.e., genesets without 
            sufficient overlap with any other geneset.")
kable_styling(tab, latex_options = "scale_down", font_size = 8)
```

In total, `r length(singletons)` singletons were identified and excluded from further analyis (Table \@ref(tab:singletons)). It is important to note that these genesets, while down-prioritized for the time being, may still be worthwile investigating later. 


```{r figure2, echo=F, fig.cap="Gene set network with singletons removed. The color scheme is the same as above. The node size corresponds to the number of genes in a set.", fig.height=4, fig.width=5}
ggnet2(net1, size = "size", max_size = 4, color = palette[V(net1)$Direction], 
       size.cut = 4, edge.size = 1, edge.color = "#99CC33")
```

```{r eval=F}
ggnet2(net1, size = "size", max_size = 4, color = palette[V(net1)$Direction], 
       size.cut = 4, edge.size = 1, edge.color = "#99CC33")
```

Figure \@ref(fig:figure2) shows the remaining network clusters with the size of the nodes representing the number of genes in the set.


### Identify binary systems (2 sets)

Next we also want to separate clusters with less than 3 gene sets. To do so, we separate disjoint subnets as individual objects, count their members, and delete all vertices belonging to clusters of size smaller than 3.

```{r binaryclusters, results="hide"}
clu1 = igraph::components(net1)
clu.lt3 = which(sizes(clu1) < 3)
v.clu.lt3 = which(clu1$membership %in% clu.lt3)
net2 = delete_vertices(net1, v.clu.lt3)
clu2 = igraph::components(net2)
in.clu.lt3 = which(dat$name %in% V(net1)$name[v.clu.lt3])
tab = dat[in.clu.lt3, ]
tab$FDR = signif(tab$FDR,2)
cludp = clu1$membership[v.clu.lt3]
cludp = data.frame(name = names(cludp), id = as.numeric(cludp))
tab = merge(tab,cludp)
tab$name = gsub("_", " ", tab$name)
tab = kable(tab[order(tab$id), c("id", "name", "NGenes", "Direction", "FDR")], 
            row.names=F, format = "latex", 
            caption = "List of binary clusters as indicated by the id column.")
kable_styling(tab, latex_options = "scale_down", font_size = 8)
```

\newpage

```{r binaries, echo=F}
kable_styling(tab, latex_options = "scale_down", font_size = 8)
```


In Table \@ref(tab:binaryclusters), consecutively listed gene sets with the same _id_ belong to the same binary cluster. Often these are gene sets from different libraries describing the same biological process or phenotype. In total, `r length(v.clu.lt3)` binary clusters were identified, for which network analysis would not be useful.


```{r figure3, echo=T, fig.cap="Gene set network with singletons and binary clusters removed. Colored according to disjoint subnetworks.", fig.height=4, fig.width=5}
set.seed(16)
nodecol = colorRampPalette(brewer.pal(9, "Set1")[sample(9)])(max(clu2$membership))
ggnet2(net2, size = "size", max_size = 4, color = nodecol[clu2$membership], 
       size.cut = 4, edge.size = 1, edge.color = "grey") 
```

Without singletons and binary clusters, we are left with larger disjoint subnets (Figure \@ref(fig:figure3)).

```{r eval=F, echo=F}
set.seed(16)
nodecol = colorRampPalette(brewer.pal(9, "Set1")[sample(9)])(max(clu2$membership))
ggnet2(net2, size = "size", max_size = 4, color = nodecol[clu2$membership], 
       size.cut = 4, edge.size = 1, edge.color = "grey") 
```


### Detect communities (sub-networks)

The larger disjoint clusters may consist of so-called _communities_, i.e., sub-networks of highly inter-connected nodes that stick together by only one or a few edges. We are using the popular _edge betweenness_ property to identify these community-connecting edges and remove them in order to split large clusters into smaller ones. 

```{r}
net2 = delete_edge_attr(net2, "weight")
clu3 = cluster_edge_betweenness(net2)
# delete edges between communities
net3 = delete_edges(net2, which(as.vector(crossing(clu3, net2))) )
# remove clusters of size <3
small_cluster_ids = which(sizes(clu3) < 3)
small_cl_v = which(clu3$membership %in% small_cluster_ids)
net3 = delete_vertices(net3, small_cl_v)

clu3 = igraph::components(net3)
nodecol = c(brewer.pal(9, "Paired"), brewer.pal(9, "Set3") )
nodecol = colorRampPalette(nodecol)(max(clu3$membership))
```

```{r figure4, echo=T, fig.cap="Disjoint clusters after community detection and splitting.", fig.height=4, fig.width=4}
ggnet2(net3, size = 0, color = nodecol[clu3$membership], 
       edge.size = 1.0, edge.color = "grey") + 
  geom_point(size = 2, color = "black") + 
  geom_point(aes(color = color), size = 1)
```

```{r eval=F, echo=F}
ggnet2(net3, size = 0, color = nodecol[clu3$membership], 
       edge.size = 1.0, edge.color = "grey") + 
  geom_point(size = 2, color="black") + 
  geom_point(aes(color = color), size = 1)
```

The result of this network-based clustering is shown in Fig. \@ref(fig:figure4)

\newpage

# Automatic annotation of gene set clusters

In analogy to the popular interactive network visualization tool _cytoscape_ [@Kucera2016], we attempt to generate automatic labels for gene set clusters. Gene set names are split into individual words and counted within each cluster. The four most frequent terms occuring at least twice are used as labels. The function `clust_head()` is defined for this purpose and contains an exclusion list of words not used.

```{r wordcount}
t.rW = c("cell", "process", "regulation", "negative", "positive", "signaling", 
         "response", "stimulus", "signal", "activity", "protein", "involved", 
         "component", "level", "effector", "event", "projection", "organismal", 
         "cellular", "modification", "pathway", "mediated", "dependent", 
         "organization", "group", "target", "biocarta", "kegg", "reactome")
clust_head = function(x){
  txt = unlist(strsplit(x, "_"))
  txt = Corpus(VectorSource(txt))
  txt = tm_map(txt, PlainTextDocument)
  txt = tm_map(txt, removePunctuation)
  txt = tm_map(txt, removeNumbers)
  txt = tm_map(txt, content_transformer(tolower))
  txt = tm_map(txt, removeWords, c(t.rW, stopwords("english")))
  tdm = TermDocumentMatrix(txt)
  m = as.matrix(tdm)
  word_freqs = sort(rowSums(m), decreasing=TRUE) 
  word_freqs = word_freqs[word_freqs>1]
  word_freqs = paste(names(word_freqs)[1:4], collapse=" ")
  gsub("[[:space:]]?NA[[:space:]]?", "", word_freqs)
}

```


### Lattice of annotated networks 

There are many possiblities to visualize geneset clusters and often a compromize between information content and crowding has to be found. Here, we are producing a lattice of network plots, one for each sub-net, with the automatic annotation as title. We begin by generating the cluster titles using the `clust_head()` function followed by cleaning up and ordering by cluster size. 

```{r}
clust = data.frame(cl = clu3$membership)
rownames(clust) = names(V(net3))
# generate cluster titles 
cl3.lab.txt = as.character(tapply(rownames(clust), clust$cl, clust_head))
# remove NAs 
cl3.lab.txt = gsub("[[:space:]]?NA[[:space:]]?", "", cl3.lab.txt)
clu3 = igraph::components(net3)
clu.order = order(clu3$csize, decreasing = T)
clu3$mem = match(clu3$membership, clu.order)
```

Then we generate a list of ggplot objects, one for each cluster or sub-net. For smaller sub-nets, the nodes are labelled with the first 4 words of their names; the first word was removed before as it is usually the name of the geneset library. For larger sub-nets, this is not feasible without overprinting. Titles are missing if none of the words from the geneset names occured more than once.

```{r}
# generate a list of ggplots
g = list(max(clu3$membership))
set.seed(7042016)
for (ii in 1:max(clu3$membership)) {
  subgf = induced_subgraph(net3, which(clu3$mem == ii))
  # generate titles with one optional line break
  title = substr(toupper(cl3.lab.txt[clu.order][ii]), 1, 60)
  if (nchar(title) > 25) {
    title = sub("(^.{10,30})[[:space:]]","\\1\\\n", title)
  }
  # generate node labels using word 2-5 of the geneset name
  v.label = names(V(subgf))
  v.label = lapply(v.label, function(x) strsplit(x, "_")[[1]])
  v.label = sapply(v.label, function(x) paste(x[2:min(5, length(x))], 
                                              collapse = "_"))
  # clean up geneset names
  v.label =  gsub("_PATHWAY","", v.label)
  v.label =  gsub("_SIGNALING", "", v.label)
  # introduce line breaks
  v.label =  gsub("_","\n", v.label)
  # remove node labels for large clusters
  if (length(v.label) > 5) v.label = rep(NA, length(v.label))
  g[[ii]] = ggnet2(subgf, edge.size = 1, edge.color = "#99CC33",
                   label = F, size=V(subgf)$size, max_size = 3,
                   size.cut = 4, color = palette[V(subgf)$Direction]) +
    theme(legend.position="none", plot.title = element_text(size=6), 
          panel.grid = element_blank()) + 
    geom_label_repel(label = v.label, size=1.2,
                     box.padding = 0.1, label.padding = 0.1) +  
    ggtitle(title) }
```


```{r eval=F}
nr.cols = min(4,max(clu3$membership))
nr.rows = ceiling(max(clu3$membership) / nr.cols)
width = sapply(g, function(x) nrow(x$data))
grid.arrange = getFromNamespace("grid.arrange", asNamespace("gridExtra"))
grid.arrange(grobs = g[seq(16)], ncol = nr.cols)
```

\newpage 

```{r figure5, echo=F, fig.width=5.5, fig.height=5.5, fig.cap="Geneset cluster with machine-generated titles. Only the first 16 connected subnets are shown. Geneset labels are omitted for clusters with more than 5 members."}
nr.cols = min(4,max(clu3$membership))
nr.rows = ceiling(max(clu3$membership) / nr.cols)
width = sapply(g, function(x) nrow(x$data))
grid.arrange = getFromNamespace("grid.arrange", asNamespace("gridExtra"))
grid.arrange(grobs = g[seq(16)], ncol = nr.cols)
```
\newpage 


\newpage 

# Discussion 

We have presented an automated workflow based on a small number of R packages for prioritization and visualization of gene set analysis results using networks, which we call RICHNET. We demonstrated how community detection facilitates categorization of differentially regulated gene sets into singletons and clusters of different size ranges. Automated label generation allowed to associate these clusters with biological themes or processes of which the member gene sets are part of.

The RICHNET workflow could be altered or extended quite naturally in a number of ways but the version presented here is the one we typically apply in our research service projects. One advantage over other approaches is that it does not depend on a particular genset library. Specific hierarchically constructed genesets, such as GO terms, would offer a straightforward way to arrive at a more global process description using higher levels in their tree structure. A second advantage is that it does not depend on the existence of a good quality gene or protein interaction network for the particular organism or disease state which is often not feasible. Only very few genesets are network-based (e.g. KEGG pathways) and would thus offer a straight-forward way to use an _i priori_ network topology. Thirdly, similar as in [@Thorsson2018], a geneset similarity network could be constructed in the form of a co-enrichment network from GSVA enrichment scores [@Hanzelmann2013] using weighted coxpression network analysis (WGCNA) [@Langfelder2008]. However, this approach relies on a relatively large sample size whereas the sample size requirement of RICHNET is not more than the GSA it relies on. 

As an alternative to the networks of genesets described here, networks of genes could be created in a reciprocal way. The underlying similarity metric between genes could be defined as the proportion of common genesets among all genesets they are part of. This approach would be equivalent to a STRING-DB network with "databases" as the only interaction allowed [@Szklarczyk2017]. 

One possible future extension of the RICHNET workflow could be the introduction of a consensus similarity metric from multiple initial networks and different community detection or cluster algorithms to improve stability against noise. A second avenue forward could be the introduction of interactive graphics in 2D or 3D [@Ognyanova2015] to allow moving, pulling, rotation or zoom and display of node specific or edge specific information. 

Some may argue in favor of encapsulating the RICHNET workflow in an R or Bioconductor package. However, it is our strong believe that for the sake of transparency and given the straightforward nature of the code it serves better to publish it openly. This way we encourage the users to addapt it to their specific requirements, to improve and expand on it.


# Data availability

The data used in this workflow is included in the _airway_ R-package[@Himes2014].


# Software availability

The R markdown file for this workflow can be [downloaded](https://gitlab.ethz.ch/nexuscbu/richnet/tree/master/RICHNET_ms/RICHNET_ms.Rmd), used and distributed according to the [Creative Commons CC BY license](https://creativecommons.org/licenses/by/4.0/).

\newpage 

```{r sessioninfo}
sessionInfo()
```


# Author contributions
MP conceptualized the content, developed the method, performed the analysis and wrote the manuscript.

# Competing interests
No competing interests were disclosed.

# Grant information
The author declared that no grants were involved in supporting this work.

# Acknowledgments
The author would like to thank all members of NEXUS and in particular Daniel Stekhoven for fruitful discussions as well as Beate Sick and Phil Cheng for critically reading the manuscript.

